pallets/corporate-actions/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+// This file is part of the Polymesh distribution (https://github.com/PolymathNetwork/Polymesh).
+// Copyright (c) 2020 Polymath
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+//! # Corporate Actions module.
+//!
+//! TODO
+//!
+//! ## Overview
+//!
+//! TODO
+//!
+//! ## Interface
+//!
+//! TODO
+//!
+//! ### Dispatchable Functions
+//!
+//! TODO
+//!
+//! ### Public Functions
+//!
+//! TODO
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(bool_to_option)]
+
+use codec::{Decode, Encode};
+use core::mem;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage,
+    dispatch::{DispatchError, DispatchResult},
+    ensure,
+    weights::Weight,
+};
+use pallet_asset as asset;
+use pallet_identity as identity;
+use polymesh_common_utilities::{
+    balances::Trait as BalancesTrait, identity::CorporateActionLink,
+    identity::Trait as IdentityTrait,
+};
+use polymesh_primitives::{AuthorizationData, IdentityId, Ticker};
+use sp_arithmetic::Permill;
+#[cfg(feature = "std")]
+use sp_runtime::{Deserialize, Serialize};
+use sp_std::prelude::*;
+
+/// How should `identities` in `TargetIdentities` be used?
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Clone, Copy, PartialEq, Eq, Encode, Decode, Debug)]
+pub enum TargetTreatment {
+    /// Only those identities should be included.
+    Include,
+    /// All identities *but* those should be included.
+    Exclude,
+}
+
+impl Default for TargetTreatment {
+    fn default() -> Self {
+        // By default, an empty list of identities to exclude means all identities are included.
+        Self::Exclude
+    }
+}
+
+/// A description of which identities that a CA will apply to.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Clone, PartialEq, Eq, Encode, Decode, Default, Debug)]
+pub struct TargetIdentities {
+    /// The specified identities either relevant or irrelevant, depending on `treatment`, for CAs.
+    pub identities: Vec<IdentityId>,
+    /// How should `identities` be treated?
+    pub treatment: TargetTreatment,
+}
+
+/// Weight abstraction for the corporate actions module.
+pub trait WeightInfo {
+    fn reset_caa() -> Weight;
+    fn set_default_targets(targets: &TargetIdentities) -> Weight;
```

Comment:
We'll have to pass `targets.identities.len()` as u32 instead of the object for the benchmarking cli to be able to parse this.

Anyway, the benchmarking CLI will generate the trait automatically when we add benchmarks. We can replace it then.